Q1. What is Min-Max scaling, and how is it used in data preprocessing? Provide an example to illustrate its
application.

Q2. What is the Unit Vector technique in feature scaling, and how does it differ from Min-Max scaling?
Provide an example to illustrate its application.

Q3. What is PCA (Principle Component Analysis), and how is it used in dimensionality reduction? Provide an
example to illustrate its application.

Q4. What is the relationship between PCA and Feature Extraction, and how can PCA be used for Feature
Extraction? Provide an example to illustrate this concept.

Q5. You are working on a project to build a recommendation system for a food delivery service. The dataset
contains features such as price, rating, and delivery time. Explain how you would use Min-Max scaling to
preprocess the data.

Q6. You are working on a project to build a model to predict stock prices. The dataset contains many
features, such as company financial data and market trends. Explain how you would use PCA to reduce the
dimensionality of the dataset.

Q7. For a dataset containing the following values: [1, 5, 10, 15, 20], perform Min-Max scaling to transform the
values to a range of -1 to 1.

Q8. For a dataset containing the following features: [height, weight, age, gender, blood pressure], perform
Feature Extraction using PCA. How many principal components would you choose to retain, and why?

In [ ]:
Q1. What is Min-Max scaling, and how is it used in data preprocessing? Provide an example to illustrate its
application.

Its a normalization technique. here all the values of the feature will be converted between zero and one


In [19]:
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
df = sns.load_dataset('tips')

In [24]:
min_max = MinMaxScaler()
df1 = pd.DataFrame(min_max.fit_transform(df[['total_bill','tip']])) 

In [26]:
df1

0         1
0    0.291579  0.001111
1    0.152283  0.073333
2    0.375786  0.277778
3    0.431713  0.256667
4    0.450775  0.290000
..        ...       ...
239  0.543779  0.546667
240  0.505027  0.111111
241  0.410557  0.111111
242  0.308965  0.083333
243  0.329074  0.222222

[244 rows x 2 columns]

In [33]:
df1[0].max()

0.9999999999999999

In [34]:
df1[0].min()

0.0

Q2. What is the Unit Vector technique in feature scaling, and how does it differ from Min-Max scaling?
Provide an example to illustrate its application.

Unit Vector Technique: Scales each feature value to have a magnitude of 1 while preserving its direction. The feature values are placed on the surface of a unit sphere.

In [35]:
from sklearn.preprocessing import normalize


In [37]:
normalize(df[['total_bill','tip']])

array([[0.99823771, 0.05934197],
       [0.98735707, 0.15851187],
       [0.98640661, 0.16432285],
       [0.99037159, 0.13843454],
       [0.98939488, 0.14525073],
       [0.98309589, 0.18309141],
       [0.97496878, 0.2223418 ],
       [0.99333102, 0.11529735],
       [0.99161511, 0.12922644],
       [0.97694312, 0.21349975],
       [0.98641987, 0.16424323],
       [0.99009498, 0.14039917],
       [0.99485672, 0.10129216],
       [0.98700924, 0.16066347],
       [0.97988851, 0.19954574],
       [0.98389908, 0.17872495],
       [0.9871829 , 0.15959298],
       [0.9750328 , 0.22206088],
       [0.97938658, 0.20199487],
       [0.98709527, 0.1601341 ],
       [0.97504726, 0.22199737],
       [0.9909398 , 0.13430677],
       [0.99014941, 0.14001479],
       [0.98201   , 0.18882891],
       [0.98737215, 0.15841793],
       [0.99147891, 0.1302673 ],
       [0.98899596, 0.14794255],
       [0.98780711, 0.15568276],
       [0.98092686, 0.19437721],
       [0.98854552, 0.15092298],
       [0.

In [ ]:
Q3. What is PCA (Principle Component Analysis), and how is it used in dimensionality reduction? Provide an
example to illustrate its application.

PCA is used to convert 2d feture or nd feature to 1d feature with a minimal loss of data.



In [38]:
import numpy as np
from sklearn.decomposition import PCA

# Sample data with two features: Height and Weight
data = np.array([[160, 55],
                 [170, 62],
                 [155, 50],
                 [175, 70],
                 [168, 60]])

# Step 1: Standardize the data
mean = np.mean(data, axis=0)
std = np.std(data, axis=0)
data_standardized = (data - mean) / std

# Step 2 and 3: Calculate Covariance Matrix and Compute Eigenvectors and Eigenvalues
cov_matrix = np.cov(data_standardized.T)
eigenvalues, eigenvectors = np.linalg.eig(cov_matrix)

# Step 4: Select Principal Components (eigenvectors with highest eigenvalues)
# In this case, we will choose the first principal component (the most important one)
top_principal_component = eigenvectors[:, 0]

# Step 5: Project Data onto New Feature Space (1D)
data_reduced = data_standardized.dot(top_principal_component)

print("Original Data:\n", data)
print("Data after Standardization:\n", data_standardized)
print("Principal Component (Top):\n", top_principal_component)
print("Data in Reduced Dimension (1D):\n", data_reduced)


Original Data:
 [[160  55]
 [170  62]
 [155  50]
 [175  70]
 [168  60]]
Data after Standardization:
 [[-0.78079594 -0.65272991]
 [ 0.61348253  0.38570404]
 [-1.47793517 -1.39446845]
 [ 1.31062176  1.5724857 ]
 [ 0.33462683  0.08900862]]
Principal Component (Top):
 [-0.70710678  0.70710678]
Data in Reduced Dimension (1D):
 [ 0.09055636 -0.16106371  0.05901989  0.18516577 -0.1736783 ]


Q4. What is the relationship between PCA and Feature Extraction, and how can PCA be used for Feature
Extraction? Provide an example to illustrate this concept.
PCA (Principal Component Analysis) is a technique that can be used for both dimensionality reduction and feature extraction. While dimensionality reduction focuses on reducing the number of dimensions (features) in a dataset, feature extraction aims to create new, more informative features from the original features.

PCA (Principal Component Analysis) is a technique that can be used for both dimensionality reduction and feature extraction. While dimensionality reduction focuses on reducing the number of dimensions (features) in a dataset, feature extraction aims to create new, more informative features from the original features.

In [39]:
import numpy as np
from sklearn.decomposition import PCA

# Sample data with two features: Height and Weight
data = np.array([[160, 55],
                 [170, 62],
                 [155, 50],
                 [175, 70],
                 [168, 60]])

# Step 1: Standardize the data
mean = np.mean(data, axis=0)
std = np.std(data, axis=0)
data_standardized = (data - mean) / std

# Step 2 and 3: Compute Principal Components
pca = PCA()
pca.fit(data_standardized)

# Step 4: Select Top Principal Components (Let's choose 1 component in this example)
k = 1
top_principal_components = pca.components_[:k]

# Step 5: Project Data onto New Feature Space (1D)
data_feature_extracted = data_standardized.dot(top_principal_components.T)

print("Original Data:\n", data)
print("Data after Standardization:\n", data_standardized)
print("Top Principal Component:\n", top_principal_components)
print("Data in Feature-Extracted Space (1D):\n", data_feature_extracted)


Original Data:
 [[160  55]
 [170  62]
 [155  50]
 [175  70]
 [168  60]]
Data after Standardization:
 [[-0.78079594 -0.65272991]
 [ 0.61348253  0.38570404]
 [-1.47793517 -1.39446845]
 [ 1.31062176  1.5724857 ]
 [ 0.33462683  0.08900862]]
Top Principal Component:
 [[0.70710678 0.70710678]]
Data in Feature-Extracted Space (1D):
 [[-1.01365585]
 [ 0.7065316 ]
 [-2.03109608]
 [ 2.03866483]
 [ 0.2995555 ]]


In [ ]:
Q5. You are working on a project to build a recommendation system for a food delivery service. The dataset
contains features such as price, rating, and delivery time. Explain how you would use Min-Max scaling to
preprocess the data.



In [41]:
import numpy as np
from sklearn.preprocessing import MinMaxScaler

data = np.array([[15, 4.5, 30],
                 [20, 4.0, 25],
                 [12, 3.8, 40],
                 [25, 4.7, 35]])

scaler = MinMaxScaler()

scaled_data = scaler.fit_transform(data)

print("Original Data:\n", data)
print("Scaled Data (Min-Max Scaling):\n", scaled_data)



Original Data:
 [[15.   4.5 30. ]
 [20.   4.  25. ]
 [12.   3.8 40. ]
 [25.   4.7 35. ]]
Scaled Data (Min-Max Scaling):
 [[0.23076923 0.77777778 0.33333333]
 [0.61538462 0.22222222 0.        ]
 [0.         0.         1.        ]
 [1.         1.         0.66666667]]


In [ ]:

Q6. You are working on a project to build a model to predict stock prices. The dataset contains many
features, such as company financial data and market trends. Explain how you would use PCA to reduce the
dimensionality of the dataset.

When working on a project to predict stock prices with a dataset that contains many features, PCA (Principal Component Analysis) can be a valuable technique for reducing the dimensionality of the dataset. By using PCA, we can extract the most relevant information from the original features and transform them into a smaller set of uncorrelated principal components. This can help simplify the model, improve computational efficiency, and potentially enhance prediction accuracy.

Here's how you can use PCA to reduce the dimensionality of the dataset:

Data Preprocessing:
Ensure that the data is properly preprocessed, handling any missing values and outliers. Additionally, it's essential to standardize the numerical features so that they have zero mean and unit variance. Standardization is crucial for PCA to treat all features equally and prevent those with larger scales from dominating the analysis.

Calculate Covariance Matrix:
Compute the covariance matrix of the standardized data. The covariance matrix represents the relationships and variances between different features.

Compute Eigenvectors and Eigenvalues:
Compute the eigenvectors and eigenvalues from the covariance matrix. The eigenvectors represent the directions (principal components) along which the data varies the most, and the eigenvalues indicate the amount of variance explained by each principal component.

Select Principal Components:
Determine the number of principal components (k) you want to retain. You can do this by looking at the explained variance ratio, which represents the proportion of variance explained by each principal component. Choose a value of k that captures a significant amount of the total variance (e.g., 95% or 99%).

Project Data onto New Feature Space:
Project the original data onto the new feature space defined by the selected k principal components. This projection will produce a new dataset with reduced dimensionality, consisting of k uncorrelated features.

Optional: Model Building and Evaluation:
Use the reduced dataset with k principal components to train your prediction model. Evaluate the model's performance and compare it to the performance when using the full set of features. Note that while PCA can improve computational efficiency and help with overfitting, it may not always result in better prediction accuracy, especially if crucial information is lost in the dimensionality reduction.

By following these steps, you can effectively use PCA to reduce the dimensionality of the dataset for predicting stock prices, transforming the original high-dimensional dataset into a smaller set of meaningful and uncorrelated features.

In [ ]:
Q7. For a dataset containing the following values: [1, 5, 10, 15, 20], perform Min-Max scaling to transform the
values to a range of -1 to 1.


In [46]:
import numpy as np
from sklearn.preprocessing import MinMaxScaler

data = np.array([[1,5,10,15,20]])
data = data.reshape(-1, 1)
scaler = MinMaxScaler()

scaled_data = scaler.fit_transform(data)

print("Original Data:\n", data)
print("Scaled Data (Min-Max Scaling):\n", scaled_data)



Original Data:
 [[ 1]
 [ 5]
 [10]
 [15]
 [20]]
Scaled Data (Min-Max Scaling):
 [[0.        ]
 [0.21052632]
 [0.47368421]
 [0.73684211]
 [1.        ]]


In [ ]:

Q8. For a dataset containing the following features: [height, weight, age, gender, blood pressure], perform
Feature Extraction using PCA. How many principal components would you choose to retain, and why?

When performing feature extraction using PCA on a dataset with features [height, weight, age, gender, blood pressure], the number of principal components to retain is a critical decision. The goal is to strike a balance between reducing dimensionality and retaining enough variance to capture the essential information.

To decide how many principal components to retain, you can consider the following approaches:

Explained Variance:
Calculate the explained variance ratio for each principal component. The explained variance ratio indicates the proportion of the total variance in the data explained by each principal component. Choose the number of principal components that collectively explain a significant portion of the total variance (e.g., 95% or 99%).

Scree Plot:
Plot the eigenvalues (variances) of the principal components in descending order. The scree plot helps visualize the proportion of variance explained by each principal component. Choose the "elbow point" on the scree plot, which is the point where the eigenvalues start to level off, indicating that additional principal components add less variance.

Domain Knowledge:
Consider the nature of the problem and the importance of specific features in the domain. If some features are known to be more critical than others, you may choose to retain fewer principal components and discard less informative features.

Let's assume you have calculated the explained variance ratio for the given dataset and obtained the following results:

Principal Component	Explained Variance Ratio
PC1	0.70
PC2	0.20
PC3	0.07
PC4	0.02
PC5	0.01
Based on the explained variance ratios, the cumulative variance explained by the first two principal components (PC1 and PC2) is 0.70 + 0.20 = 0.90. This means that these two principal components capture 90% of the total variance in the data. You may decide to retain these two principal components as they explain a significant portion of the information in the original features.

Ultimately, the choice of the number of principal components to retain depends on the specific requirements of your project, the amount of variance you want to preserve, and the trade-off between dimensionality reduction and information loss. In this case, retaining the first two principal components (PC1 and PC2) seems like a reasonable choice as they collectively capture a high proportion of the data's variance and can effectively represent the essential information in the original features.